### PR TITLE
Enhance Bearer Auth

### PR DESCRIPTION
**Pull request by GregorEeckels**
BearerAuthenticationReader now requires a Lambda function upon instantiation which will be used to create the output for the readAuthentication call.
The idea for this is based on the fact that one might not wish to load a PreAuthenticatedAuthenticationToken into the SecurityContext as the authentication provider might not support that particular token. Hence with the lambda function one can use any kind of object as long as it is a subtype of Authentication.

Also the BearerAuth example was expanded to be closer to a real world use case.
In this case an instance of Keycloak ([link](https://www.keycloak.org/)) is used.

Could we get this into a RELEASE version asap? Thank you.

### CODE REVIEW

File: examples/security-grpc-bearerAuth-server/src/main/java/net/devh/boot/grpc/examples/security/server/SecurityConfiguration.java [hunk 1]
```diff
@@ -67,7 +68,10 @@ AuthenticationManager authenticationManager() {
     @Bean
     // Configure which authentication types you support.
     GrpcAuthenticationReader authenticationReader() {
-        return new BearerAuthenticationReader();
+        return new BearerAuthenticationReader(
+                accessToken -> new PreAuthenticatedAuthenticationToken(accessToken, null));
+        // Alternatively
+        // return new BearerAuthenticationReader(accessToken -> new BearerTokenAuthenticationToken(accessToken));
```
[the remark this answers is not in the document]
Comment (GregorEeckels) [reply examples/security-grpc-bearerAuth-server/src/main/java/net/devh/boot/grpc/examples/security/server/SecurityConfiguration.java]: Done

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/authentication/BearerAuthenticationReader.java [hunk 2]
```diff
@@ -20,15 +20,17 @@
...
-import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken;
 
 import io.grpc.Metadata;
 import io.grpc.ServerCall;
 import lombok.extern.slf4j.Slf4j;
 
 /**
- * Reads {@link PreAuthenticatedAuthenticationToken bearer token} from the request.
+ * Reads bearer token from the request. Upon constructing the Reader, one can define the output object of type
+ * Authentication which will be return by readAuthentication.
```
Comment (ST-DDT) [inline grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/authentication/BearerAuthenticationReader.java]: The later sentence is a little bit hard to understand. Maybe rephrase it like this:

````java
Spring-Security has several token-based {@link AuthenticationProvider} implementations (e.g. in spring-security-web/oauth2 or spring-security-oauth2-resource-server), so you need to provide a {@link Function} that wraps the extracted token in a {@link Authentication} object supported by your AuthenticationProvider.
````

and maybe also show an example usage:

````java
 * <p>
 * <b>Example-Usage:</b>
 * </p>
 *
 * For spring-security-web:
 *
 * <pre>
 * <code>new BearerAuthenticationReader(token -&gt; new PreAuthenticatedAuthenticationToken(token, null))</code>
 * </pre>
 *
 * For spring-security-oauth2-resource-server:
 *
 * <pre>
 * <code>new BearerAuthenticationReader(token -&gt; new BearerTokenAuthenticationToken(token))</code>
 * </pre>
````

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/authentication/BearerAuthenticationReader.java [hunk 3]
```diff
@@ -38,6 +40,12 @@
     private static final String PREFIX = BEARER_AUTH_PREFIX.toLowerCase();
     private static final int PREFIX_LENGTH = PREFIX.length();
 
+    private Function<String, Authentication> tokenWrapper;
+
+    public BearerAuthenticationReader(Function<String, Authentication> tokenWrapper) {
```
Comment (ST-DDT) [inline grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/authentication/BearerAuthenticationReader.java]: Please add some javadocs to the method that describe what will be passed to the function.

Maybe like this:

````java
/**
 * Creates a new BearerAuthenticationReader with the given wrapper function.
 * 
 * @param tokenWrapper The function used to convert the token (without bearer prefix) into an
 *        {@link Authentication} object.
 */
````

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/authentication/BearerAuthenticationReader.java [hunk 4]
```diff
@@ -38,6 +40,12 @@
...
+        this.tokenWrapper = tokenWrapper;
```
Comment (ST-DDT) [inline grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/security/authentication/BearerAuthenticationReader.java]: Also add an null check here

File: examples/security-grpc-bearerAuth-server/src/main/java/net/devh/boot/grpc/examples/security/server/DevAuthenticationProvider.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+
+public class DevAuthenticationProvider implements AuthenticationProvider {
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        return new Authentication() {
```
Comment (ST-DDT) [inline examples/security-grpc-bearerAuth-server/src/main/java/net/devh/boot/grpc/examples/security/server/DevAuthenticationProvider.java]: I would recommend using an existing `Authentication` class.
`return new UsernamePasswordToken("dev-auth","", Arrays.asList(new SimpleGrantedAuthority("ROLE_TEST")));`

File: examples/security-grpc-bearerAuth-server/src/main/java/net/devh/boot/grpc/examples/security/server/DummyAuthenticationProvider.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @author Gregor Eeckels (gregor.eeckels@gmail.com)
+ */
+public class DummyAuthenticationProvider implements AuthenticationProvider {
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        return new UsernamePasswordAuthenticationToken("dev-auth", "",
+                // Add required ROLES here
+                Arrays.asList(new SimpleGrantedAuthority("ROLE_TEST")));
```
Comment (ST-DDT) [inline examples/security-grpc-bearerAuth-server/src/main/java/net/devh/boot/grpc/examples/security/server/DummyAuthenticationProvider.java]: Just found this spring util: `AuthorityUtils.createAuthorityList(String...)`

**Comment by GregorEeckels** [comment]
Ready to be merged

**Comment by ST-DDT** [comment]
I squashed the commits to remove the many formatting commits.
